### PR TITLE
Improved namespace renaming for XML/XSD

### LINKED ISSUE #366: Ability to rename a namespace.
If the main namespace name is highlighted, and rename is called on it, replace all instances of it in the element names.

eg:

```xml
<root xmlns:xs="...">
<xs:element />
<xs:complexType />
```

```xml
<root xmlns:xs="...">  <!-- highlight 'xs' and renamed to 'xsd' -->
<xs:element />
<xs:complexType />
```

```xml
<root xmlns:xsd="...">
<xsd:element />
<xsd:complexType />
```

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #366

Can rename namespace declaration and all will change, renaming element with namespace does not wipe
namespace, attribute values referencing a namespace will update if necessary

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java [hunk 1]
```diff
@@ -32,4 +32,6 @@
 	 */
 	CMElementDeclaration findCMElement(DOMElement element, String namespace);
 
+	String getURI();
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java]: Should ne not in an another pr?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 2]
```diff
@@ -101,9 +112,26 @@ private void fillWithChildrenElementDeclaration(DOMElement element, Collection<C
 			CompletionItem item = new CompletionItem(label);
 			item.setFilterText(request.getFilterForStartTagName(label));
 			item.setKind(CompletionItemKind.Property);
-			String documentation = child.getDocumentation();
-			if (documentation != null) {
-				item.setDetail(documentation);
+			StringBuilder sb = new StringBuilder();
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: I think this code should ne set on the cm document implementaion. I suggest you that you create an abstractcmdocument class which provides the getfulldocumentation method. This abstract class could store the uri too

Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: @angelozerr 

CMDTDDocument is already extending a class, while CMXSDDocument is not.

CMDTDElementDeclaration is also extending a class, while CMXSDElementDeclaration is not.

Is there something I could do to avoid this problem?

Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: Your right @NikolasKomonen please forget my comment

Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: A another suggestion is to set the documentation code with source in a static method in XMLGenerator. Set the build of documentation with source will help more how this doc is built.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java [hunk 3]
```diff
@@ -91,12 +91,12 @@ public XMLHighlighting(XMLExtensionsRegistry extensionsRegistry) {
...
 	}
 
-	private static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
+	public static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
 		return startTagRange != null && covers(startTagRange, position)
 				|| endTagRange != null && covers(endTagRange, position);
 	}
 
-	private static List<DocumentHighlight> getHighlightsList(Range startTagRange, Range endTagRange) {
+	public static List<DocumentHighlight> getHighlightsList(Range startTagRange, Range endTagRange) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java]: why update to public?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java [hunk 4]
```diff
@@ -91,12 +91,12 @@ public XMLHighlighting(XMLExtensionsRegistry extensionsRegistry) {
 		return Collections.emptyList();
 	}
 
-	private static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
+	public static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java]: I think it should be better to move this static method to an utility class like XMLPositionUtility

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java [hunk 5]
```diff
@@ -113,11 +113,11 @@ private static boolean isBeforeOrEqual(Position pos1, Position pos2) {
 				|| (pos1.getLine() == pos2.getLine() && pos1.getCharacter() <= pos2.getCharacter());
 	}
 
-	private static boolean covers(Range range, Position position) {
+	public static boolean covers(Range range, Position position) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java]: I think it should be better to move this static method to an utility class like XMLPositionUtility

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java [hunk 6]
```diff
@@ -113,11 +113,11 @@ private static boolean isBeforeOrEqual(Position pos1, Position pos2) {
 				|| (pos1.getLine() == pos2.getLine() && pos1.getCharacter() <= pos2.getCharacter());
 	}
 
-	private static boolean covers(Range range, Position position) {
+	public static boolean covers(Range range, Position position) {
 		return isBeforeOrEqual(range.getStart(), position) && isBeforeOrEqual(position, range.getEnd());
 	}
 
-	private static Range getTagNameRange(TokenType tokenType, int startOffset, DOMDocument xmlDocument) {
+	public static Range getTagNameRange(TokenType tokenType, int startOffset, DOMDocument xmlDocument) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLHighlighting.java]: I think it should be better to move this static method to an utility class like XMLPositionUtility

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4xml.commons.BadLocationException;
+import org.eclipse.lsp4xml.dom.DOMAttr;
+import org.eclipse.lsp4xml.dom.DOMDocument;
+import org.eclipse.lsp4xml.dom.DOMElement;
+import org.eclipse.lsp4xml.dom.DOMNode;
+import org.eclipse.lsp4xml.dom.parser.TokenType;
+
+/**
+ * XMLRename
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Explain more what is the goal of this class like rename namespaces and add @author

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+ */
+public class XMLRename {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: XMLRename is too generic I think. Perhaps XMLRenameNamespace should be better?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+			try {
+				startPos = xmlDocument.positionAt(node.getStart());
+				endPos = xmlDocument.positionAt(node.getEnd());
+				tempRange = new Range(startPos, endPos);
+
+			} catch (BadLocationException e) {
+				LOGGER.log(Level.SEVERE, "In XMLRename the Node at provided Offset is a BadLocation", e);
+				return Collections.emptyList();
+			}
+			if (XMLHighlighting.covers(tempRange, position)) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: use import static

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java (hunk 7 above)
Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Before for the Rename feature, it was using XMLHighlighting to get Highlights and then it turned them into TextEdits. Also Highlighting would only work for 1 Element (Start Tag and End Tag) but with XMLRename it will handle many elements with the same name, namespaces and maybe other renaming in the future.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java (hunk 8 above)
Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: You think so? We have XMLFormatter, XMLHighlighting, XMLCompletions ... 

This class will do all renaming, not only namespace.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLLanguageService.java [hunk 10]
```diff
@@ -164,8 +162,7 @@ private static void publishOneDiagnosticInRoot(DOMDocument document, String mess
 	}
 
 	public WorkspaceEdit doRename(DOMDocument xmlDocument, Position position, String newText) {
-		List<TextEdit> textEdits = findDocumentHighlights(xmlDocument, position).stream()
-				.map(h -> new TextEdit(h.getRange(), newText)).collect(Collectors.toList());
+		List<TextEdit> textEdits = XMLRename.getTextEdits(xmlDocument, position, newText);
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLLanguageService.java]: Please follow the same instance pattern that other features. In other word XMLRename#getTextEdits should be not static and the service should have a XMLRename rename field.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java (hunk 8 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Thank's for the clarification, I understand more now. Please add the 2 rename features (rename tag element and rename namespaces by using ui, li in the javadoc)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+				List<DOMAttr> attributes = element.getAttributeNodes();
+
+				if(attributes == null) {
+					return Collections.emptyList();
+				}
+
+				for (DOMAttr attr : attributes) {
+					DOMNode nameNode = attr.getNodeAttrName();
+
+					if(!attr.getName().startsWith("xmlns")) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: I think it should b emore clean to have a DOMAttr#isXmlns() method. You can find isXmlNs in the DOMElement which should be moved to DOMAttr

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+				edits.add(new TextEdit(new Range(start, end), newNamespace));
+			}
+		}
+
+		//Renames all elements with oldNamespace
+		List<DOMNode> children = Arrays.asList(rootElement);
+		return renameElementsNamespace(document, edits, children, oldNamespace, newNamespace);
+	}
+
+	public static List<TextEdit> renameElementsNamespace(DOMDocument document, List<TextEdit> edits, List<DOMNode> elements, String oldNamespace, String newNamespace) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: why public?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+					renameElementsNamespace(document, edits, element.getChildren(), oldNamespace, newNamespace);
+				}
+			}
+		}
+		
+		return edits;
+		
+	}
+
+	public static List<TextEdit> renameElementNamespace(DOMDocument document, DOMElement element, int oldNamespaceLength, String newNamespace) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: why public?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java (hunk 7 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Ok thanks for the clarification

**Comment by NikolasKomonen** [comment]
@angelozerr Updated, moved the methods to DOMAttr. Implemented doRename() to follow the pattern for XMLLanguageService

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java [hunk 14]
```diff
@@ -528,4 +535,39 @@ public static Range getElementDeclMissingContentOrCategory(int offset, DOMDocume
 		return null;
 	}
 
+	public static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java [hunk 15]
```diff
@@ -528,4 +535,39 @@ public static Range getElementDeclMissingContentOrCategory(int offset, DOMDocume
 		return null;
 	}
 
+	public static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
+		return startTagRange != null && covers(startTagRange, position)
+				|| endTagRange != null && covers(endTagRange, position);
+	}
+
+	public static boolean covers(Range range, Position position) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java [hunk 16]
```diff
@@ -528,4 +535,39 @@ public static Range getElementDeclMissingContentOrCategory(int offset, DOMDocume
...
+	public static boolean doesTagCoverPosition(Range startTagRange, Range endTagRange, Position position) {
+		return startTagRange != null && covers(startTagRange, position)
+				|| endTagRange != null && covers(endTagRange, position);
+	}
+
+	public static boolean covers(Range range, Position position) {
+		return isBeforeOrEqual(range.getStart(), position) && isBeforeOrEqual(position, range.getEnd());
+	}
+
+	public static boolean isBeforeOrEqual(Position pos1, Position pos2) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java [hunk 17]
```diff
@@ -528,4 +535,39 @@ public static Range getElementDeclMissingContentOrCategory(int offset, DOMDocume
...
+		return pos1.getLine() < pos2.getLine()
+				|| (pos1.getLine() == pos2.getLine() && pos1.getCharacter() <= pos2.getCharacter());
+	}
+
+	public static Range getTagNameRange(TokenType tokenType, int startOffset, DOMDocument xmlDocument) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/XMLPositionUtility.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java [hunk 18]
```diff
@@ -38,9 +38,9 @@ public void testFilesCachePathPreference() throws Exception {
 	@Test
 	public void normalizePathTest() {
 		assertEquals(Paths.get(System.getProperty("user.home"), "Test", "Folder").toString(), FilesUtils.normalizePath("~/Test/Folder"));
-		assertEquals(Paths.get(separator, "Test", "~", "Folder").toString(), FilesUtils.normalizePath("/Test/~/Folder"));
+		assertEquals(Paths.get(separator + "Test", "~", "Folder").toString(), FilesUtils.normalizePath("/Test/~/Folder"));
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java]: It should belong to another PR?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+					renameElementsNamespace(document, edits, element.getChildren(), oldNamespace, newNamespace);
+				}
+			}
+		}
+		
+		return edits;
+		
+	}
+
+	private static List<TextEdit> renameElementNamespace(DOMDocument document, DOMElement element, int oldNamespaceLength, String newNamespace) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+		}
+		for (Range r : ranges) {
+			if(r != null) {
+				edits.add(new TextEdit(r, newNamespace));
+			}
+		}
+		return edits;
+	}
+
+	private static List<TextEdit> renameElementAttributeValueNamespace(DOMDocument document, DOMElement element, String oldNamespace, String newNamespace) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+						}
+						edits.add(new TextEdit(new Range(start, end), newNamespace));
+					}
+				}
+			}
+		}
+		return edits;
+	}
+
+	private static Range[] createNamespaceRange(DOMDocument document, DOMElement element, int namespaceLength) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+				if (startTagRange != null) {
+					result.add(new TextEdit(startTagRange, newText));
+				}
+				if (endTagRange != null) {
+					result.add(new TextEdit(endTagRange, newText));
+				}
+				return result;
+	}
+
+	private static List<TextEdit> renameAllNamespaceOccurrences(DOMDocument document, String oldNamespace, String newNamespace, @Nullable DOMAttr rootAttr) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java [hunk 23]
```diff
@@ -0,0 +1,299 @@
...
+						return renameAllNamespaceOccurrences(xmlDocument, namespaceName, newText, attr);
+					}
+				}
+			}
+		}
+
+		return Collections.emptyList();
+	}
+
+	private static List<TextEdit> getRenameList(Range startTagRange, Range endTagRange, String newText) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLRename.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMElement.java [hunk 24]
```diff
@@ -10,6 +10,9 @@
  */
 package org.eclipse.lsp4xml.dom;
 
+import static org.eclipse.lsp4xml.dom.DOMAttr.XMLNS_ATTR;
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMElement.java]: Keep the XMLNS_ATTR constants as private and provide a static method DOMAttr#isXmlns(String attributeName) which can be used in DOMAttr#isXmlNs and in DOMElement

**Comment by angelozerr** [comment]
@NikolasKomonen I played with your PR and it seems it works great I will continue to play with it again).

The rename of namespace prefix is working when you do a rename to the xmlns: attribute (I mean xmlns of xmlns:qq):

```xml
<aa:a xmlns:aa="aa.com" xmlns:qq="qq.com">
	<aa:b></aa:b>
	<aa:b></aa:b>
	<aa:c />
	<t type="aa:b" />
	<qq:b></qq:b>
</aa:a>
```

It should be very cool if we could rename the aa of an element (I mean <aa by an another name)

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java (hunk 18 above)
Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java]: It should be, but there was a problem with git history in master. This will fix the problem.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMElement.java (hunk 24 above)
Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMElement.java]: It is hard to make it private, DOMElement needs this value: https://github.com/angelozerr/lsp4xml/blob/master/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMElement.java#L123

But I did changes and made static isXmlns(String) and other methods too.

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java (hunk 18 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/utils/FilesUtilsTest.java]: ok no pb
